File: languages/o/ocaml/lambda_core.ml
(* LOGIC *)
let _true = fun x -> fun _y -> x
let _false = fun _x -> fun y -> y
let _not = fun b -> b _false _true
let _and = fun b1 -> fun b2 -> b1 b2 _false
let _or = fun b1 -> fun b2 -> b1 _true b2

(* CHURCH NUMERALS *)
let _zero = fun _f -> fun x -> x
let _succ = fun n -> fun f -> fun x -> f (n f x)

let _pred =
  fun n -> fun f -> fun x -> n (fun g -> fun h -> h (g f)) (fun _u -> x) (fun a -> a)
;;

let _one = _succ _zero
let read_bool = fun b -> b "true" "false" |> print_endline
let read_church = fun n -> n (fun x -> x + 1) 0 |> string_of_int |> print_endline

let () =
  (* EXAMPLE *)
  print_endline "LOGIC";
  print_endline "----------------";
  print_endline "TRUE/FALSE";
  read_bool _true;
  read_bool _false;
  print_endline "NOT";
  read_bool (_not _true);
  read_bool (_not _false);
  print_endline "AND";
  read_bool (_and _true _true);
  read_bool (_and _true _false);
  read_bool (_and _false _true);
  read_bool (_and _false _false);
  print_endline "OR";
  read_bool (_or _true _true);
  read_bool (_or _true _false);
  read_bool (_or _false _true);
  read_bool (_or _false _false);
  print_endline "\nCHURCH NUMERALS";
  print_endline "----------------";
  print_endline "ZERO/SUCC";
  read_church _zero;
  read_church _one;
  read_church (_succ _one);
  read_church (_succ (_succ _one));
  print_endline "PRED";
  read_church (_pred (_succ _zero));
  read_church (_pred (_succ (_succ _zero)));
  read_church (_pred (_succ (_succ (_succ _zero))))
;;

File: languages/g/go/lambda-core.go
package main

import "fmt"

type T func(t T) T

// LOGIC
var (
	_true  T = func(x T) T { return func(y T) T { return x } }
	_false T = func(x T) T { return func(y T) T { return y } }
	_not   T = func(b T) T { return b(_false)(_true) }
	_and   T = func(b1 T) T { return func(b2 T) T { return b1(b2)(_false) } }
	_or    T = func(b1 T) T { return func(b2 T) T { return b1(_true)(b2) } }
)

// CHURCH NUMERALS
var (
	_zero T = func(f T) T { return func(x T) T { return x } }
	_succ T = func(n T) T {
		return func(f T) T {
			return func(x T) T {
				return f(n(f)(x))
			}
		}
	}
	_pred T = func(n T) T {
		return func(f T) T {
			return func(x T) T {
				return n(func(g T) T { return func(h T) T { return h(g(f)) } })(func(u T) T { return x })(func(a T) T { return a })
			}
		}
	}
	_one T = _succ(_zero)
)

// HELPERS - not pure lambda calculus
func readBool(b T) bool {
	var res bool
	var closureTrue T = func(t T) T {
		res = true
		return t
	}
	var closureFalse T = func(t T) T {
		res = false
		return t
	}
	b(closureTrue)(closureFalse)(func(t T) T { return t })
	return res
}

func readChurch(n T) int {
	res := 0
	var closurePlusOne T = func(t T) T {
		res++
		return t
	}
	n(closurePlusOne)(func(t T) T { return t })
	return res
}

func main() {
	fmt.Println("LOGIC")
	fmt.Println("-------------")
	fmt.Println("TRUE/FALSE")
	fmt.Println(readBool(_true))
	fmt.Println(readBool(_false))

	fmt.Println("NOT")
	fmt.Println(readBool(_not(_true)))
	fmt.Println(readBool(_not(_false)))

	fmt.Println("AND")
	fmt.Println(readBool(_and(_true)(_true)))
	fmt.Println(readBool(_and(_true)(_false)))
	fmt.Println(readBool(_and(_false)(_true)))
	fmt.Println(readBool(_and(_false)(_false)))

	fmt.Println("OR")
	fmt.Println(readBool(_or(_true)(_true)))
	fmt.Println(readBool(_or(_true)(_false)))
	fmt.Println(readBool(_or(_false)(_true)))
	fmt.Println(readBool(_or(_false)(_false)))

	fmt.Println("\nCHURCH NUMERALS")
	fmt.Println("-------------")
	fmt.Println("ZERO/SUCC")
	fmt.Println(readChurch(_zero))
	fmt.Println(readChurch(_one))
	fmt.Println(readChurch(_succ(_one)))
	fmt.Println(readChurch(_succ(_succ(_one))))
	fmt.Println("PRED")
	fmt.Println(readChurch(_pred(_one)))
	fmt.Println(readChurch(_pred(_succ(_one))))
	fmt.Println(readChurch(_pred(_succ(_succ(_one)))))

}
